File: tools/iKeyd.ProfileCompiler/IKeydBehaviorDsl.cs
using System.Text.Json;
using System.Text.Json.Nodes;
using System.Text.RegularExpressions;
using iKeyd.Core.Chords;

internal sealed record IKeydBehaviorDslExtension(string CleanSource, JsonObject Layers, JsonObject Behaviors);

internal static class IKeydBehaviorDsl
{
    private const string Ident = "[A-Za-z0-9_]+";
    private const string KeyRef = Ident + @"(?:\[\s*\d+\s*,\s*\d+\s*\]|\." + Ident + @")?";

    private static readonly string[][] Jis109Layout =
    [
        ["Escape", "F1", "F2", "F3", "F4", "F5", "F6", "F7", "F8", "F9", "F10", "F11", "F12", "PrintScreen", "ScrollLock", "Pause"],
        ["ZenkakuHankaku", "1", "2", "3", "4", "5", "6", "7", "8", "9", "0", "Minus", "Caret", "Yen", "Backspace"],
        ["Tab", "Q", "W", "E", "R", "T", "Y", "U", "I", "O", "P", "AT", "LeftBracket"],
        ["CapsLock", "A", "S", "D", "F", "G", "H", "J", "K", "L", "SColon", "Colon", "RightBracket", "Enter"],
        ["LeftShift", "Z", "X", "C", "V", "B", "N", "M", "Comma", "Dot", "Slash", "Ro", "RightShift"],
        ["LeftControl", "LeftGui", "LeftAlt", "Muhenkan", "Space", "Henkan", "KatakanaHiragana", "RightAlt", "RightGui", "Menu", "RightControl"],
        ["Insert", "Home", "PageUp"],
        ["Delete", "End", "PageDown"],
        ["Left", "Up", "Down", "Right"],
        ["NumLock", "NumpadSlash", "NumpadAsterisk", "NumpadMinus"],
        ["Numpad7", "Numpad8", "Numpad9", "NumpadPlus"],
        ["Numpad4", "Numpad5", "Numpad6"],
        ["Numpad1", "Numpad2", "Numpad3", "NumpadEnter"],
        ["Numpad0", "NumpadDot"]
    ];

    public static IKeydBehaviorDslExtension Extract(string source, string path)
    {
        ArgumentNullException.ThrowIfNull(source);
        var normalized = source.Replace("\r\n", "\n", StringComparison.Ordinal).Replace('\r', '\n');
        var lines = normalized.Split('\n');
        var layouts = ScanLayouts(lines, path);
        var clean = lines.ToArray();
        var layers = new JsonObject();
        var behaviors = new JsonObject();

        string? extractedKind = null;
        string? extractedName = null;
        string? behaviorTrigger = null;
        JsonObject? behaviorDraft = null;
        var ordinaryDepth = 0;

        for (var index = 0; index < lines.Length; index++)
        {
            var lineNumber = index + 1;
            var line = StripComment(lines[index]).Trim();
            if (line.Length == 0)
                continue;

            if (extractedKind is not null)
            {
                clean[index] = string.Empty;
                if (line == "}")
                {
                    if (extractedKind == "behavior")
                    {
                        ValidateBehaviorDraft(path, lineNumber, behaviorTrigger!, behaviorDraft!, layers);
                        behaviors.Add(behaviorTrigger!, behaviorDraft);
                    }

                    extractedKind = null;
                    extractedName = null;
                    behaviorTrigger = null;
                    behaviorDraft = null;
                    continue;
                }

                if (extractedKind == "layer")
                {
                    var match = Regex.Match(line, $@"^({KeyRef})\s*=\s*(.+)$", RegexOptions.CultureInvariant);
                    if (!match.Success)
                        throw Error(path, lineNumber, $"unknown layer statement: {line}");
                    var key = ResolveAndValidateKey(path, lineNumber, match.Groups[1].Value, layouts);
                    var action = ParseAction(path, lineNumber, match.Groups[2].Value, allowHoldActions: false);
                    var layer = (JsonObject)layers[extractedName!]!;
                    if (ContainsPropertyIgnoreCase(layer, key))
                        throw Error(path, lineNumber, $"duplicate key '{key}' in behavior layer '{extractedName}'");
                    layer.Add(key, action);
                    continue;
                }

                ParseBehaviorSetting(path, lineNumber, line, behaviorDraft!);
                continue;
            }

            if (ordinaryDepth > 0)
            {
                if (line.EndsWith("{", StringComparison.Ordinal))
                    ordinaryDepth++;
                if (line == "}")
                    ordinaryDepth--;
                continue;
            }

            if (IsOrdinaryTopLevelBlock(line))
            {
                ordinaryDepth = 1;
                continue;
            }

            var layerStart = Regex.Match(line, $@"^layer\s+({Ident})\s*\{{$", RegexOptions.CultureInvariant);
            if (layerStart.Success)
            {
                var name = layerStart.Groups[1].Value;
                if (ContainsPropertyIgnoreCase(layers, name))
                    throw Error(path, lineNumber, $"duplicate behavior layer '{name}'");
                layers.Add(name, new JsonObject());
                extractedKind = "layer";
                extractedName = name;
                clean[index] = string.Empty;
                continue;
            }

            var behaviorBlock = Regex.Match(line, $@"^behavior\s+({KeyRef})\s*\{{$", RegexOptions.CultureInvariant);
            if (behaviorBlock.Success)
            {
                var trigger = ResolveAndValidateKey(path, lineNumber, behaviorBlock.Groups[1].Value, layouts);
                if (ContainsPropertyIgnoreCase(behaviors, trigger))
                    throw Error(path, lineNumber, $"duplicate behavior trigger '{trigger}'");
                extractedKind = "behavior";
                behaviorTrigger = trigger;
                behaviorDraft = NewBehaviorDraft();
                clean[index] = string.Empty;
                continue;
            }

            var behaviorShort = Regex.Match(line, $@"^behavior\s+({KeyRef})\s*=\s*(.+)$", RegexOptions.CultureInvariant);
            if (behaviorShort.Success)
            {
                var trigger = ResolveAndValidateKey(path, lineNumber, behaviorShort.Groups[1].Value, layouts);
                if (ContainsPropertyIgnoreCase(behaviors, trigger))
                    throw Error(path, lineNumber, $"duplicate behavior trigger '{trigger}'");
                var behavior = ParseShorthand(path, lineNumber, behaviorShort.Groups[2].Value);
                ValidateBehaviorDraft(path, lineNumber, trigger, behavior, layers, validateLayerReference: false);
                behaviors.Add(trigger, behavior);
                clean[index] = string.Empty;
                continue;
            }
        }

        if (extractedKind is not null)
            throw Error(path, lines.Length, $"unclosed {extractedKind} block");

        foreach (var property in behaviors)
            ValidateBehaviorDraft(path, 1, property.Key, (JsonObject)property.Value!, layers);

        return new IKeydBehaviorDslExtension(string.Join('\n', clean), layers, behaviors);
    }

    public static string Merge(string baseJson, IKeydBehaviorDslExtension extension)
    {
        var root = JsonNode.Parse(baseJson) as JsonObject
            ?? throw new InvalidDataException("DSL compiler produced a non-object profile.");
        if (extension.Layers.Count > 0)
            root["layers"] = extension.Layers.DeepClone();
        if (extension.Behaviors.Count > 0)
            root["behaviors"] = extension.Behaviors.DeepClone();
        return root.ToJsonString();
    }

    private static Dictionary<string, List<List<string>>> ScanLayouts(string[] lines, string path)
    {
        var layouts = new Dictionary<string, List<List<string>>>(StringComparer.OrdinalIgnoreCase);
        string? current = null;

        for (var index = 0; index < lines.Length; index++)
        {
            var lineNumber = index + 1;
            var line = StripComment(lines[index]).Trim();
            if (line.Length == 0)
                continue;

            if (current is not null)
            {
                if (line == "}")
                {
                    current = null;
                    continue;
                }
                var row = Regex.Match(line, @"^row\s+(.+)$", RegexOptions.CultureInvariant);
                if (row.Success)
                    layouts[current].Add(ParseLayoutRow(path, lineNumber, row.Groups[1].Value));
                continue;
            }

            var keyboard = Regex.Match(line, $@"^keyboard\s+({Ident})\s*;?$", RegexOptions.CultureInvariant);
            if (keyboard.Success && string.Equals(keyboard.Groups[1].Value, "JIS109", StringComparison.OrdinalIgnoreCase))
            {
                layouts["JIS109"] = Jis109Layout.Select(row => row.ToList()).ToList();
                continue;
            }

            var layout = Regex.Match(line, $@"^layout\s+({Ident})\s*\{{$", RegexOptions.CultureInvariant);
            if (layout.Success)
            {
                current = layout.Groups[1].Value;
                layouts.TryAdd(current, []);
            }
        }

        return layouts;
    }

    private static List<string> ParseLayoutRow(string path, int lineNumber, string value)
    {
        var keys = Regex.Split(value.Trim().TrimEnd(';'), @"[\s,]+")
            .Where(item => item.Length > 0)
            .ToList();
        if (keys.Count == 0 || keys.Any(key => !Regex.IsMatch(key, $@"^{Ident}$", RegexOptions.CultureInvariant)))
            throw Error(path, lineNumber, "expected one or more key identifiers after 'row'");
        return keys;
    }

    private static JsonObject ParseShorthand(string path, int lineNumber, string expression)
    {
        expression = expression.Trim().TrimEnd(';').Trim();
        var layerTap = Regex.Match(expression, $@"^layer_tap\(\s*({Ident})\s*,\s*({Ident})\s*\)$", RegexOptions.IgnoreCase | RegexOptions.CultureInvariant);
        if (layerTap.Success)
        {
            ValidateOutputKey(path, lineNumber, layerTap.Groups[2].Value);
            var result = NewBehaviorDraft();
            result["tap"] = Action("key", CanonicalKey(layerTap.Groups[2].Value));
            result["hold"] = Action("layer", layerTap.Groups[1].Value);
            return result;
        }

        var modTap = Regex.Match(expression, $@"^mod_tap\(\s*({Ident})\s*,\s*({Ident})\s*\)$", RegexOptions.IgnoreCase | RegexOptions.CultureInvariant);
        if (modTap.Success)
        {
            ValidateOutputKey(path, lineNumber, modTap.Groups[2].Value);
            var result = NewBehaviorDraft();
            result["tap"] = Action("key", CanonicalKey(modTap.Groups[2].Value));
            result["hold"] = Action("modifier", NormalizeModifier(path, lineNumber, modTap.Groups[1].Value));
            return result;
        }

        var hold = ParseAction(path, lineNumber, expression, allowHoldActions: true);
        var kind = hold["kind"]!.GetValue<string>();
        if (kind is not ("layer" or "modifier"))
            throw Error(path, lineNumber, "behavior shorthand must be layer_tap, mod_tap, layer(...) or modifier(...)");
        var behavior = NewBehaviorDraft();
        behavior["hold"] = hold;
        return behavior;
    }

    private static JsonObject NewBehaviorDraft() => new()
    {
        ["timeoutMs"] = 180,
        ["interrupt"] = "hold"
    };

    private static void ParseBehaviorSetting(string path, int lineNumber, string line, JsonObject draft)
    {
        var assignment = Regex.Match(line, $@"^({Ident})\s*=\s*(.+)$", RegexOptions.CultureInvariant);
        if (!assignment.Success)
            throw Error(path, lineNumber, $"unknown behavior setting: {line}");

        var name = assignment.Groups[1].Value.ToLowerInvariant();
        var value = assignment.Groups[2].Value.Trim().TrimEnd(';').Trim();
        switch (name)
        {
            case "tap":
                if (draft.ContainsKey("tap"))
                    throw Error(path, lineNumber, "duplicate behavior tap setting");
                draft["tap"] = ParseAction(path, lineNumber, value, allowHoldActions: false);
                break;
            case "hold":
                if (draft.ContainsKey("hold"))
                    throw Error(path, lineNumber, "duplicate behavior hold setting");
                var hold = ParseAction(path, lineNumber, value, allowHoldActions: true);
                if (hold["kind"]!.GetValue<string>() is not ("layer" or "modifier"))
                    throw Error(path, lineNumber, "hold must be layer(...) or modifier(...)");
                draft["hold"] = hold;
                break;
            case "timeout":
                var timeout = Regex.Match(value, @"^(\d+)\s*ms$", RegexOptions.IgnoreCase | RegexOptions.CultureInvariant);
                if (!timeout.Success || !int.TryParse(timeout.Groups[1].Value, out var timeoutMs) || timeoutMs <= 0)
                    throw Error(path, lineNumber, "timeout must be a positive duration such as 180ms");
                draft["timeoutMs"] = timeoutMs;
                break;
            case "interrupt":
                var policy = value.ToLowerInvariant();
                if (policy is not ("hold" or "tap"))
                    throw Error(path, lineNumber, "interrupt must be 'hold' or 'tap'");
                draft["interrupt"] = policy;
                break;
            default:
                throw Error(path, lineNumber, $"unknown behavior setting '{assignment.Groups[1].Value}'");
        }
    }

    private static JsonObject ParseAction(string path, int lineNumber, string expression, bool allowHoldActions)
    {
        expression = expression.Trim().TrimEnd(';').Trim();
        var call = Regex.Match(expression, $@"^({Ident})\((.*)\)$", RegexOptions.CultureInvariant);
        if (!call.Success)
            throw Error(path, lineNumber, $"invalid behavior action '{expression}'");

        var kind = call.Groups[1].Value.ToLowerInvariant();
        var argument = call.Groups[2].Value.Trim();
        switch (kind)
        {
            case "key":
                if (!Regex.IsMatch(argument, $@"^{Ident}$", RegexOptions.CultureInvariant))
                    throw Error(path, lineNumber, "key(...) expects one key name");
                ValidateOutputKey(path, lineNumber, argument);
                return Action("key", CanonicalKey(argument));
            case "text":
                try
                {
                    var text = JsonSerializer.Deserialize<string>(argument)
                        ?? throw Error(path, lineNumber, "text(...) requires a quoted string");
                    return Action("text", text);
                }
                catch (JsonException exception)
                {
                    throw Error(path, lineNumber, $"text(...) requires a quoted string: {exception.Message}");
                }
            case "layer" when allowHoldActions:
                if (!Regex.IsMatch(argument, $@"^{Ident}$", RegexOptions.CultureInvariant))
                    throw Error(path, lineNumber, "layer(...) expects one layer name");
                return Action("layer", argument);
            case "modifier" when allowHoldActions:
                return Action("modifier", NormalizeModifier(path, lineNumber, argument));
            default:
                throw Error(path, lineNumber, allowHoldActions
                    ? "action must be key(...), text(...), layer(...) or modifier(...)"
                    : "action must be key(...) or text(...)");
        }
    }

    private static JsonObject Action(string kind, string value) => new()
    {
        ["kind"] = kind,
        ["value"] = value
    };

    private static void ValidateBehaviorDraft(
        string path,
        int lineNumber,
        string trigger,
        JsonObject draft,
        JsonObject layers,
        bool validateLayerReference = true)
    {
        if (draft["hold"] is not JsonObject hold)
            throw Error(path, lineNumber, $"behavior '{trigger}' requires hold = layer(...) or modifier(...)");
        if (draft["tap"] is JsonObject tap && tap["kind"]!.GetValue<string>() is "layer" or "modifier")
            throw Error(path, lineNumber, $"behavior '{trigger}' tap action must be key or text");

        if (validateLayerReference && hold["kind"]!.GetValue<string>() == "layer")
        {
            var layerName = hold["value"]!.GetValue<string>();
            if (!ContainsPropertyIgnoreCase(layers, layerName))
                throw Error(path, lineNumber, $"behavior '{trigger}' references unknown layer '{layerName}'");
        }
    }

    private static string ResolveAndValidateKey(
        string path,
        int lineNumber,
        string keyRef,
        Dictionary<string, List<List<string>>> layouts)
    {
        var key = ResolveKeyRef(path, lineNumber, keyRef, layouts);
        if (!KeyId.TryParseCompact(key, out var code))
            throw Error(path, lineNumber, $"behavior key '{key}' is not a supported physical key");
        return new KeyId(code).Value;
    }

    private static string ResolveKeyRef(
        string path,
        int lineNumber,
        string value,
        Dictionary<string, List<List<string>>> layouts)
    {
        value = value.Trim();
        if (Regex.IsMatch(value, $@"^{Ident}$", RegexOptions.CultureInvariant))
            return value;

        var named = Regex.Match(value, $@"^({Ident})\.({Ident})$", RegexOptions.CultureInvariant);
        if (named.Success)
        {
            var layoutName = ResolveLayoutName(named.Groups[1].Value, layouts);
            if (layoutName is null || !layouts.TryGetValue(layoutName, out var layout))
                throw Error(path, lineNumber, $"unknown layout '{named.Groups[1].Value}' in key reference '{value}'");
            var requested = named.Groups[2].Value;
            foreach (var key in layout.SelectMany(row => row))
                if (string.Equals(key, requested, StringComparison.OrdinalIgnoreCase))
                    return key;
            throw Error(path, lineNumber, $"layout '{named.Groups[1].Value}' has no key named '{requested}'");
        }

        var coordinate = Regex.Match(value, $@"^({Ident})\[\s*(\d+)\s*,\s*(\d+)\s*\]$", RegexOptions.CultureInvariant);
        if (!coordinate.Success)
            throw Error(path, lineNumber, $"invalid key reference '{value}'");
        var resolvedName = ResolveLayoutName(coordinate.Groups[1].Value, layouts);
        if (resolvedName is null || !layouts.TryGetValue(resolvedName, out var resolved))
            throw Error(path, lineNumber, $"unknown layout '{coordinate.Groups[1].Value}' in key reference '{value}'");
        var row = int.Parse(coordinate.Groups[2].Value, System.Globalization.CultureInfo.InvariantCulture);
        var column = int.Parse(coordinate.Groups[3].Value, System.Globalization.CultureInfo.InvariantCulture);
        if (row < 1 || row > resolved.Count)
            throw Error(path, lineNumber, $"row {row} is out of range for layout '{coordinate.Groups[1].Value}'");
        if (column < 1 || column > resolved[row - 1].Count)
            throw Error(path, lineNumber, $"column {column} is out of range for layout '{coordinate.Groups[1].Value}' row {row}");
        return resolved[row - 1][column - 1];
    }

    private static string? ResolveLayoutName(string requested, Dictionary<string, List<List<string>>> layouts)
    {
        if (!string.Equals(requested, "POS", StringComparison.OrdinalIgnoreCase))
            return layouts.Keys.FirstOrDefault(name => string.Equals(name, requested, StringComparison.OrdinalIgnoreCase));
        return layouts.ContainsKey("POS") ? "POS"
            : layouts.ContainsKey("JIS109") ? "JIS109"
            : layouts.ContainsKey("BASE") ? "BASE"
            : null;
    }

    private static void ValidateOutputKey(string path, int lineNumber, string key)
    {
        if (!KeyId.TryParseCompact(key, out _))
            throw Error(path, lineNumber, $"unknown output key '{key}'");
    }

    private static string CanonicalKey(string key)
    {
        KeyId.TryParseCompact(key, out var code);
        return new KeyId(code).Value;
    }

    private static string NormalizeModifier(string path, int lineNumber, string value)
        => value.Trim().ToLowerInvariant() switch
        {
            "ctrl" or "control" => "Control",
            "shift" => "Shift",
            "alt" => "Alt",
            "gui" or "win" or "super" => "Gui",
            _ => throw Error(path, lineNumber, $"unknown modifier '{value.Trim()}'")
        };

    private static bool IsOrdinaryTopLevelBlock(string line)
        => Regex.IsMatch(line, $@"^(profile\s+{Ident}|layout\s+{Ident}|keymap\s+{Ident}(?:\s+using\s+{Ident})?|quirks)\s*\{{$", RegexOptions.CultureInvariant);

    private static bool ContainsPropertyIgnoreCase(JsonObject obj, string name)
        => obj.Any(property => string.Equals(property.Key, name, StringComparison.OrdinalIgnoreCase));

    private static string StripComment(string line)
    {
        var inString = false;
        var escaped = false;
        for (var index = 0; index < line.Length; index++)
        {
            var character = line[index];
            if (inString)
            {
                if (escaped) escaped = false;
                else if (character == '\\') escaped = true;
                else if (character == '"') inString = false;
                continue;
            }
            if (character == '"') inString = true;
            else if (character == '/' && index + 1 < line.Length && line[index + 1] == '/') return line[..index];
        }
        return line;
    }

    private static Exception Error(string path, int lineNumber, string message)
        => new IKeydDslException(path, lineNumber, message);
}

File: tools/ikeyd_behavior_dsl.py
from __future__ import annotations

import json
import re
from collections import OrderedDict
from pathlib import Path
from typing import Any


def _strip_comment(line: str) -> str:
    in_string = False
    escaped = False
    for index, ch in enumerate(line):
        if in_string:
            if escaped:
                escaped = False
            elif ch == "\\":
                escaped = True
            elif ch == '"':
                in_string = False
            continue
        if ch == '"':
            in_string = True
        elif ch == "/" and index + 1 < len(line) and line[index + 1] == "/":
            return line[:index]
    return line


def _compact_key_map(base: Any) -> dict[str, str]:
    keys = [key for row in base.JIS109_LAYOUT for key in row] + ["NumpadComma"]
    aliases = {
        "esc": "Escape",
        "return": "Enter",
        "spc": "Space",
        "bs": "Backspace",
        "bspc": "Backspace",
        "pgup": "PageUp",
        "pgdn": "PageDown",
        "lctrl": "LeftControl",
        "rctrl": "RightControl",
        "lshift": "LeftShift",
        "rshift": "RightShift",
        "lalt": "LeftAlt",
        "ralt": "RightAlt",
        "lgui": "LeftGui",
        "rgui": "RightGui",
        "lwin": "LeftGui",
        "rwin": "RightGui",
        "semicolon": "SColon",
        "period": "Dot",
    }
    result = {key.casefold(): key for key in keys}
    result.update(aliases)
    return result


def _canonical_key(base: Any, path: Path, lineno: int, value: str) -> str:
    mapping = _compact_key_map(base)
    canonical = mapping.get(value.casefold())
    if canonical is None:
        raise base.DslError(path, lineno, f"unknown behavior key '{value}'")
    return canonical


def _scan_layouts(base: Any, lines: list[str], path: Path) -> OrderedDict[str, list[list[str]]]:
    layouts: OrderedDict[str, list[list[str]]] = OrderedDict()
    current: str | None = None
    for lineno, raw in enumerate(lines, 1):
        line = _strip_comment(raw).strip()
        if not line:
            continue
        if current is not None:
            if line == "}":
                current = None
                continue
            row = re.fullmatch(r"row\s+(.+)", line)
            if row:
                layouts[current].append(base.parse_layout_row(path, lineno, row.group(1)))
            continue
        keyboard = re.fullmatch(rf"keyboard\s+({base.IDENT})\s*;?", line)
        if keyboard and keyboard.group(1).casefold() == "jis109":
            layouts["JIS109"] = [row.copy() for row in base.JIS109_LAYOUT]
            continue
        layout = re.fullmatch(rf"layout\s+({base.IDENT})\s*\{{", line)
        if layout:
            current = layout.group(1)
            layouts.setdefault(current, [])
    return layouts


def _action(base: Any, path: Path, lineno: int, expression: str, *, allow_hold: bool) -> OrderedDict[str, str]:
    expression = expression.strip().rstrip(";").strip()
    call = re.fullmatch(rf"({base.IDENT})\((.*)\)", expression)
    if not call:
        raise base.DslError(path, lineno, f"invalid behavior action '{expression}'")
    kind = call.group(1).casefold()
    arg = call.group(2).strip()
    if kind == "key":
        if not re.fullmatch(base.IDENT, arg):
            raise base.DslError(path, lineno, "key(...) expects one key name")
        return OrderedDict(kind="key", value=_canonical_key(base, path, lineno, arg))
    if kind == "text":
        try:
            value = json.loads(arg)
        except json.JSONDecodeError as exc:
            raise base.DslError(path, lineno, f"text(...) requires a quoted string: {exc.msg}") from exc
        if not isinstance(value, str):
            raise base.DslError(path, lineno, "text(...) requires a quoted string")
        return OrderedDict(kind="text", value=value)
    if kind == "layer" and allow_hold:
        if not re.fullmatch(base.IDENT, arg):
            raise base.DslError(path, lineno, "layer(...) expects one layer name")
        return OrderedDict(kind="layer", value=arg)
    if kind == "modifier" and allow_hold:
        modifiers = {
            "ctrl": "Control", "control": "Control", "shift": "Shift",
            "alt": "Alt", "gui": "Gui", "win": "Gui", "super": "Gui",
        }
        value = modifiers.get(arg.casefold())
        if value is None:
            raise base.DslError(path, lineno, f"unknown modifier '{arg}'")
        return OrderedDict(kind="modifier", value=value)
    allowed = "key(...), text(...), layer(...) or modifier(...)" if allow_hold else "key(...) or text(...)"
    raise base.DslError(path, lineno, f"action must be {allowed}")


def _new_behavior() -> OrderedDict[str, object]:
    return OrderedDict(timeoutMs=180, interrupt="hold")


def _shorthand(base: Any, path: Path, lineno: int, expression: str) -> OrderedDict[str, object]:
    expression = expression.strip().rstrip(";").strip()
    layer_tap = re.fullmatch(rf"layer_tap\(\s*({base.IDENT})\s*,\s*({base.IDENT})\s*\)", expression, re.I)
    if layer_tap:
        result = _new_behavior()
        result["tap"] = OrderedDict(kind="key", value=_canonical_key(base, path, lineno, layer_tap.group(2)))
        result["hold"] = OrderedDict(kind="layer", value=layer_tap.group(1))
        return result
    mod_tap = re.fullmatch(rf"mod_tap\(\s*({base.IDENT})\s*,\s*({base.IDENT})\s*\)", expression, re.I)
    if mod_tap:
        result = _new_behavior()
        result["tap"] = OrderedDict(kind="key", value=_canonical_key(base, path, lineno, mod_tap.group(2)))
        result["hold"] = _action(base, path, lineno, f"modifier({mod_tap.group(1)})", allow_hold=True)
        return result
    hold = _action(base, path, lineno, expression, allow_hold=True)
    if hold["kind"] not in {"layer", "modifier"}:
        raise base.DslError(path, lineno, "behavior shorthand must be layer_tap, mod_tap, layer(...) or modifier(...)")
    result = _new_behavior()
    result["hold"] = hold
    return result


def _validate_behavior(base: Any, path: Path, lineno: int, trigger: str, behavior: OrderedDict[str, object], layers: OrderedDict[str, object], *, check_layer: bool = True) -> None:
    hold = behavior.get("hold")
    if not isinstance(hold, dict) or hold.get("kind") not in {"layer", "modifier"}:
        raise base.DslError(path, lineno, f"behavior '{trigger}' requires hold = layer(...) or modifier(...)")
    tap = behavior.get("tap")
    if isinstance(tap, dict) and tap.get("kind") not in {"key", "text"}:
        raise base.DslError(path, lineno, f"behavior '{trigger}' tap action must be key or text")
    if check_layer and hold.get("kind") == "layer":
        name = str(hold.get("value"))
        if not any(existing.casefold() == name.casefold() for existing in layers):
            raise base.DslError(path, lineno, f"behavior '{trigger}' references unknown layer '{name}'")


def extract(base: Any, text: str, path: Path) -> tuple[str, OrderedDict[str, OrderedDict[str, object]], OrderedDict[str, OrderedDict[str, object]]]:
    lines = text.splitlines()
    layouts = _scan_layouts(base, lines, path)
    clean = lines.copy()
    layers: OrderedDict[str, OrderedDict[str, object]] = OrderedDict()
    behaviors: OrderedDict[str, OrderedDict[str, object]] = OrderedDict()
    extracted: str | None = None
    extracted_name: str | None = None
    behavior_trigger: str | None = None
    behavior_draft: OrderedDict[str, object] | None = None
    ordinary_depth = 0

    for index, raw in enumerate(lines):
        lineno = index + 1
        line = _strip_comment(raw).strip()
        if not line:
            continue

        if extracted is not None:
            clean[index] = ""
            if line == "}":
                if extracted == "behavior":
                    assert behavior_trigger is not None and behavior_draft is not None
                    _validate_behavior(base, path, lineno, behavior_trigger, behavior_draft, layers)
                    behaviors[behavior_trigger] = behavior_draft
                extracted = extracted_name = behavior_trigger = None
                behavior_draft = None
                continue
            if extracted == "layer":
                match = re.fullmatch(rf"({base.KEY_REF})\s*=\s*(.+)", line)
                if not match:
                    raise base.DslError(path, lineno, f"unknown layer statement: {line}")
                key = _canonical_key(base, path, lineno, base.resolve_key_ref(path, lineno, match.group(1), layouts))
                layer = layers[extracted_name or ""]
                if any(existing.casefold() == key.casefold() for existing in layer):
                    raise base.DslError(path, lineno, f"duplicate key '{key}' in behavior layer '{extracted_name}'")
                layer[key] = _action(base, path, lineno, match.group(2), allow_hold=False)
                continue

            assert behavior_draft is not None
            assignment = re.fullmatch(rf"({base.IDENT})\s*=\s*(.+)", line)
            if not assignment:
                raise base.DslError(path, lineno, f"unknown behavior setting: {line}")
            name = assignment.group(1).casefold()
            value = assignment.group(2).strip().rstrip(";").strip()
            if name == "tap":
                if "tap" in behavior_draft:
                    raise base.DslError(path, lineno, "duplicate behavior tap setting")
                behavior_draft["tap"] = _action(base, path, lineno, value, allow_hold=False)
            elif name == "hold":
                if "hold" in behavior_draft:
                    raise base.DslError(path, lineno, "duplicate behavior hold setting")
                hold = _action(base, path, lineno, value, allow_hold=True)
                if hold["kind"] not in {"layer", "modifier"}:
                    raise base.DslError(path, lineno, "hold must be layer(...) or modifier(...)")
                behavior_draft["hold"] = hold
            elif name == "timeout":
                timeout = re.fullmatch(r"(\d+)\s*ms", value, re.I)
                if not timeout or int(timeout.group(1)) <= 0:
                    raise base.DslError(path, lineno, "timeout must be a positive duration such as 180ms")
                behavior_draft["timeoutMs"] = int(timeout.group(1))
            elif name == "interrupt":
                policy = value.casefold()
                if policy not in {"hold", "tap"}:
                    raise base.DslError(path, lineno, "interrupt must be 'hold' or 'tap'")
                behavior_draft["interrupt"] = policy
            else:
                raise base.DslError(path, lineno, f"unknown behavior setting '{assignment.group(1)}'")
            continue

        if ordinary_depth:
            if line.endswith("{"):
                ordinary_depth += 1
            if line == "}":
                ordinary_depth -= 1
            continue

        if re.fullmatch(rf"(?:profile\s+{base.IDENT}|layout\s+{base.IDENT}|keymap\s+{base.IDENT}(?:\s+using\s+{base.IDENT})?|quirks)\s*\{{", line):
            ordinary_depth = 1
            continue

        layer_start = re.fullmatch(rf"layer\s+({base.IDENT})\s*\{{", line)
        if layer_start:
            name = layer_start.group(1)
            if any(existing.casefold() == name.casefold() for existing in layers):
                raise base.DslError(path, lineno, f"duplicate behavior layer '{name}'")
            layers[name] = OrderedDict()
            extracted, extracted_name = "layer", name
            clean[index] = ""
            continue

        block = re.fullmatch(rf"behavior\s+({base.KEY_REF})\s*\{{", line)
        if block:
            trigger = _canonical_key(base, path, lineno, base.resolve_key_ref(path, lineno, block.group(1), layouts))
            if any(existing.casefold() == trigger.casefold() for existing in behaviors):
                raise base.DslError(path, lineno, f"duplicate behavior trigger '{trigger}'")
            extracted, behavior_trigger, behavior_draft = "behavior", trigger, _new_behavior()
            clean[index] = ""
            continue

        short = re.fullmatch(rf"behavior\s+({base.KEY_REF})\s*=\s*(.+)", line)
        if short:
            trigger = _canonical_key(base, path, lineno, base.resolve_key_ref(path, lineno, short.group(1), layouts))
            if any(existing.casefold() == trigger.casefold() for existing in behaviors):
                raise base.DslError(path, lineno, f"duplicate behavior trigger '{trigger}'")
            behavior = _shorthand(base, path, lineno, short.group(2))
            _validate_behavior(base, path, lineno, trigger, behavior, layers, check_layer=False)
            behaviors[trigger] = behavior
            clean[index] = ""

    if extracted is not None:
        raise base.DslError(path, len(lines), f"unclosed {extracted} block")
    for trigger, behavior in behaviors.items():
        _validate_behavior(base, path, 1, trigger, behavior, layers)
    return "\n".join(clean), layers, behaviors


def merge(profile: dict[str, object], layers: OrderedDict[str, object], behaviors: OrderedDict[str, object]) -> OrderedDict[str, object]:
    result: OrderedDict[str, object] = OrderedDict()
    for key, value in profile.items():
        if key == "knownQuirks":
            if layers:
                result["layers"] = layers
            if behaviors:
                result["behaviors"] = behaviors
        result[key] = value
    if "knownQuirks" not in profile:
        if layers:
            result["layers"] = layers
        if behaviors:
            result["behaviors"] = behaviors
    return result
